api/src/permissions/modules/validate-access/lib/validate-item-access.ts

```diff
@@ -43,22 +55,63 @@ export async function validateItemAccess(options: ValidateItemAccessOptions, con
 		},
 	};
 
+	// Inject the root fields after the permissions have been processed, as to not require access to all collection fields
+	if (options.returnAllowedRootFields) {
+		const policies = await fetchPolicies(options.accountability, context);
+
+		const permissions = await fetchPermissions(
+			{ action: options.action, policies, collections: [options.collection], accountability: options.accountability },
+			context,
+		);
+
+		const schemaFields = Object.keys(context.schema.collections[options.collection]!.fields);
+		const hasWildcard = permissions.some((p) => p.fields?.includes('*'));
+
+		const allowedFields = hasWildcard ? new Set(schemaFields) : new Set(permissions.flatMap((p) => p.fields ?? []));
+
+		// Create children only for fields that exist in schema and are allowed by permissions
+		ast.children = Array.from(allowedFields)
```

Comment:
In the interest of reducing code duplication let's replace this with `fetchAllowedFields`
```suggestion
		const allowedFields = await fetchAllowedFields(
			{ action: options.action, collections: [options.collection], accountability: options.accountability },
			context,
		);

		const schemaFields = Object.keys(context.schema.collections[options.collection]!.fields);
		const hasWildcard = allowedFields.some((field) => field === '*');

		const permissionedFields = hasWildcard ? schemaFields : fetchAllowedFields;

		// Create children only for fields that exist in schema and are allowed by permissions
		ast.children = Array.from(permissionedFields)
```

Reply:
Good catch ! 
Used `fetchAllowedFields` as suggested. 
Note: I still kept `fetchPolicies/fetchPermissions` for injectCases() below since it needs the full permissions object, but since they're cached there's no performance impact

Reply:
> Note: I still kept fetchPolicies/fetchPermissions for injectCases() below since it needs the full permissions object, but since they're cached there's no performance impact

That is unfortunate :(

Reply:
Ideally they would be optionally passed in, otherwise defaulted to fetching from cache. Something we can look into in a different PR if needed.